File: go/src/leetcode/isPalindrome/isPalindrome.go
package main

import "fmt"

/**
* Definition for singly-linked list.
*/

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	var slow= head
	var fast= head
	var nh *ListNode
	if head == nil || head.Next == nil {
		return true
	}
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		head.Next = nh
		nh = head
		head = slow
	}
	if fast != nil {
		slow = slow.Next
	}
	head = nh
	for head != nil {
		if head.Val != slow.Val {
			return false
		}
		head = head.Next
		slow = slow.Next
	}
	return true
}

func print(head *ListNode) {
	for head != nil {
		fmt.Printf("%v ", head.Val)
		head = head.Next
	}
	fmt.Println()
}

func main() {
	var node1 = &ListNode{1, nil}
	var node2 = &ListNode{2, nil}
	var node3 = &ListNode{2, nil}
	var node4 = &ListNode{1, nil}
	node1.Next = node2
	node2.Next = node3
	node3.Next = node4
	print(node1)
	r := isPalindrome(node1)
	fmt.Println(r)
}

File: python-pycharm/isPalindrome.py
#! -*- coding:utf-8 -*-

# 编写一个函数，检查输入的链表是否是回文的。
# 示例 1：
#
# 输入： 1->2
# 输出： false
# 示例 2：
#
# 输入： 1->2->2->1
# 输出： true
#  
# 进阶：
# 你能否用 O(n) 时间复杂度和 O(1) 空间复杂度解决此题？

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return True
        slow = head
        fast = head
        new_head = None
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            head.next = new_head
            new_head = head
            head = slow

        if fast:
            slow = slow.next
        head = new_head
        while head:
            if head.val != slow.val:
                return False
            head = head.next
            slow = slow.next
        return True


def print_list(node_list):
    while node_list:
        # print(node_list.val, end=" ")
        print node_list.val,
        node_list = node_list.next
    print("")


if __name__ == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node3 = ListNode(2)
    node4 = ListNode(1)
    node1.next = node2
    node2.next = node3
    node3.next = node4
    print_list(node1)

    s = Solution()
    r = s.isPalindrome(node1)
    print(r)
